File: src/backend/core/contracts/crawl.py
"""
Acceptance Criteria:
- CrawlJobCreate requires pipeline_id (inferred from route), source_connector_id (optional),
  job_type, trigger, max_attempts, rate_limit_per_minute (optional),
  robots_txt_url (optional), concurrency_budget (optional).
- CrawlJobResponse exposes all CrawlJob fields.
- CrawlJobUpdate allows updating status, error_code, error_message, trace_id.
- CrawlArtifactResponse exposes all CrawlArtifact fields.
- CrawlBudgetCreate requires client_id, pipeline_id (optional), source_connector_id (optional),
  adapter_key (optional), budget_type, max_concurrent.
- CrawlBudgetResponse exposes all CrawlBudget fields.
- ArtifactStoreRequest carries url, content, mime_type, artifact_type, policy_decision,
  crawl_job_id (optional), source_connector_id (optional), seed_lead_row_id (optional).
- RobotsCheckResult carries url, allowed (bool), reason (optional).
- No TypeScript `any`.
"""

import uuid
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict

ArtifactType = Literal[
    "html_page", "search_result", "profile_evidence",
    "provider_response", "import_artifact",
]

CrawlJobStatus = Literal[
    "queued", "running", "paused", "failed", "retrying",
    "completed", "blocked", "stale", "cancelled",
]


class CrawlJobCreate(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    source_connector_id: uuid.UUID | None = None
    job_type: str = "crawl"
    trigger: str = "api"
    max_attempts: int = 3
    rate_limit_per_minute: int | None = None
    robots_txt_url: str | None = None
    concurrency_budget: int | None = None


class CrawlJobUpdate(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    status: CrawlJobStatus | None = None
    error_code: str | None = None
    error_message: str | None = None
    trace_id: str | None = None
    started_at: datetime | None = None
    finished_at: datetime | None = None


class CrawlJobResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True, populate_by_name=True)

    id: uuid.UUID
    client_id: uuid.UUID
    pipeline_id: uuid.UUID
    source_connector_id: uuid.UUID | None
    job_type: str
    status: str
    idempotency_key: str
    trigger: str
    attempt: int
    max_attempts: int
    lease_owner: str | None
    lease_expires_at: datetime | None
    heartbeat_at: datetime | None
    scheduled_at: datetime
    started_at: datetime | None
    finished_at: datetime | None
    error_code: str | None
    error_message: str | None
    trace_id: str | None
    rate_limit_per_minute: int | None
    robots_txt_url: str | None
    concurrency_budget: int | None
    created_at: datetime
    updated_at: datetime


class CrawlArtifactResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True, populate_by_name=True)

    id: uuid.UUID
    client_id: uuid.UUID
    pipeline_id: uuid.UUID
    crawl_job_id: uuid.UUID | None
    source_connector_id: uuid.UUID | None
    seed_lead_row_id: uuid.UUID | None
    artifact_type: str
    url: str | None
    storage_key: str
    content_hash: str | None
    status: str
    policy_decision: str | None
    mime_type: str | None
    size_bytes: int | None
    raw_metadata_json: str | None
    created_at: datetime
    updated_at: datetime


class CrawlBudgetCreate(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    pipeline_id: uuid.UUID | None = None
    source_connector_id: uuid.UUID | None = None
    adapter_key: str | None = None
    budget_type: str
    max_concurrent: int = 1


class CrawlBudgetResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True, populate_by_name=True)

    id: uuid.UUID
    client_id: uuid.UUID
    pipeline_id: uuid.UUID | None
    source_connector_id: uuid.UUID | None
    adapter_key: str | None
    budget_type: str
    max_concurrent: int
    current_count: int
    created_at: datetime
    updated_at: datetime


class ArtifactStoreRequest(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    url: str | None = None
    content: str
    mime_type: str = "text/html"
    artifact_type: ArtifactType = "html_page"
    policy_decision: str | None = None
    crawl_job_id: uuid.UUID | None = None
    source_connector_id: uuid.UUID | None = None
    seed_lead_row_id: uuid.UUID | None = None


class RobotsCheckResult(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    url: str
    allowed: bool
    reason: str | None = None

File: src/frontend/web/src/contracts/crawl.ts
/**
 * Acceptance Criteria:
 * - CrawlJobResponseSchema validates crawl job API responses.
 * - CrawlArtifactResponseSchema validates artifact responses.
 * - CrawlBudgetResponseSchema validates budget responses.
 * - RobotsCheckResultSchema validates robots check responses.
 * - All schemas use strict Zod types; no TypeScript `any`.
 */
import { z } from "zod";

export const CrawlJobStatusSchema = z.enum([
  "queued",
  "running",
  "paused",
  "failed",
  "retrying",
  "completed",
  "blocked",
  "stale",
  "cancelled",
]);

export const ArtifactTypeSchema = z.enum([
  "html_page",
  "search_result",
  "profile_evidence",
  "provider_response",
  "import_artifact",
]);

export const CrawlJobResponseSchema = z.object({
  id: z.string().min(1),
  client_id: z.string().min(1),
  pipeline_id: z.string().min(1),
  source_connector_id: z.string().nullable(),
  job_type: z.string(),
  status: CrawlJobStatusSchema,
  idempotency_key: z.string(),
  trigger: z.string(),
  attempt: z.number(),
  max_attempts: z.number(),
  lease_owner: z.string().nullable(),
  lease_expires_at: z.string().nullable(),
  heartbeat_at: z.string().nullable(),
  scheduled_at: z.string(),
  started_at: z.string().nullable(),
  finished_at: z.string().nullable(),
  error_code: z.string().nullable(),
  error_message: z.string().nullable(),
  trace_id: z.string().nullable(),
  rate_limit_per_minute: z.number().nullable(),
  robots_txt_url: z.string().nullable(),
  concurrency_budget: z.number().nullable(),
  created_at: z.string(),
  updated_at: z.string(),
});

export const CrawlJobCreateSchema = z.object({
  source_connector_id: z.string().optional(),
  job_type: z.string().optional(),
  trigger: z.string().optional(),
  max_attempts: z.number().optional(),
  rate_limit_per_minute: z.number().optional(),
  robots_txt_url: z.string().optional(),
  concurrency_budget: z.number().optional(),
});

export const CrawlArtifactResponseSchema = z.object({
  id: z.string().min(1),
  client_id: z.string().min(1),
  pipeline_id: z.string().min(1),
  crawl_job_id: z.string().nullable(),
  source_connector_id: z.string().nullable(),
  seed_lead_row_id: z.string().nullable(),
  artifact_type: ArtifactTypeSchema,
  url: z.string().nullable(),
  storage_key: z.string(),
  content_hash: z.string().nullable(),
  status: z.string(),
  policy_decision: z.string().nullable(),
  mime_type: z.string().nullable(),
  size_bytes: z.number().nullable(),
  raw_metadata_json: z.string().nullable(),
  created_at: z.string(),
  updated_at: z.string(),
});

export const ArtifactStoreRequestSchema = z.object({
  url: z.string().optional(),
  content: z.string().min(1),
  mime_type: z.string().optional(),
  artifact_type: ArtifactTypeSchema.optional(),
  policy_decision: z.string().optional(),
  crawl_job_id: z.string().optional(),
  source_connector_id: z.string().optional(),
  seed_lead_row_id: z.string().optional(),
});

export const CrawlBudgetResponseSchema = z.object({
  id: z.string().min(1),
  client_id: z.string().min(1),
  pipeline_id: z.string().nullable(),
  source_connector_id: z.string().nullable(),
  adapter_key: z.string().nullable(),
  budget_type: z.string(),
  max_concurrent: z.number(),
  current_count: z.number(),
  created_at: z.string(),
  updated_at: z.string(),
});

export const RobotsCheckResultSchema = z.object({
  url: z.string(),
  allowed: z.boolean(),
  reason: z.string().nullable(),
});

export type CrawlJobResponse = z.infer<typeof CrawlJobResponseSchema>;
export type CrawlJobCreate = z.infer<typeof CrawlJobCreateSchema>;
export type CrawlArtifactResponse = z.infer<typeof CrawlArtifactResponseSchema>;
export type ArtifactStoreRequest = z.infer<typeof ArtifactStoreRequestSchema>;
export type CrawlBudgetResponse = z.infer<typeof CrawlBudgetResponseSchema>;
export type RobotsCheckResult = z.infer<typeof RobotsCheckResultSchema>;

export const CrawlJobListResponseSchema = z.array(CrawlJobResponseSchema);
export const CrawlArtifactListResponseSchema = z.array(CrawlArtifactResponseSchema);
export const CrawlBudgetListResponseSchema = z.array(CrawlBudgetResponseSchema);
